tools/canmonitor.cpp

```diff
@@ -20,21 +20,83 @@ using AS::CAN::KvaserCanUtils;
 using AS::CAN::KvaserChannel;
 using AS::CAN::KvaserCan;
 using AS::CAN::ReturnStatuses;
+using AS::CAN::CanMsg;
 
 KvaserCan kv_can;
+uint32_t channel_idx = 0;
+uint32_t bitrate = 500000;
 
-void shutdown()
+void shutdown_with_error(const ReturnStatuses &ret, const int &error_num)
 {
+  std::cerr << static_cast<int>(ret);
+  std::cerr << " - " << KvaserCanUtils::returnStatusDesc(ret).c_str();
+  std::cerr << std::endl;
   kv_can.close();
+  exit(error_num);
 }
 
 void sig_handler(int s)
 {
   std::cout << "Caught signal " << s << std::endl;
-  shutdown();
+  kv_can.close();
   exit(1);
 }
 
+void can_read()
+{
+  // Read CAN data
+  auto ret = ReturnStatuses::OK;
+
+  ret = kv_can.open(channel_idx, bitrate);
```

Comment:
The open() function checks for on_bus status and just returns OK if it's already open. However, you did remind me that I need to be running close() at the end of the can_read() function. can_read() is the callback and it will loop internally (the while(true) below) until all messages in the buffer have been read, which also clears the event buffer so another callback will trigger. However, we have seen issues where, if you leave the channel open between subsequent read loops, it sometimes fails to read. I'll fix that as well, which will make this call necessary on subsequent callbacks.

Reply:
Added the close() call outside of the while(true) loop.